File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/parser.mjs
import { parse as parseYaml } from "yaml";

export class MissingFrontmatterError extends Error {
  constructor(message) {
    super(message);
    this.name = "MissingFrontmatterError";
  }
}

export class MalformedFrontmatterError extends Error {
  constructor(message, cause) {
    super(message);
    this.name = "MalformedFrontmatterError";
    this.cause = cause;
  }
}

const DEP_RE =
  /^(?:(?<owner>[A-Za-z0-9_.-]+)\/(?<repo>[A-Za-z0-9_.-]+)#)?(?<name>[a-z0-9-]+)(?:@(?<range>.+))?$/;

function normalizeDep(entry) {
  if (entry == null) return null;
  if (typeof entry === "string") {
    const m = DEP_RE.exec(entry.trim());
    if (!m) {
      return {
        name: entry.trim(),
        version_range: null,
        origin: "in_repo",
        owner: null,
        repo: null,
      };
    }
    const { owner, repo, name, range } = m.groups;
    return {
      name,
      version_range: range ?? null,
      origin: owner ? "cross_repo" : "in_repo",
      owner: owner ?? null,
      repo: repo ?? null,
    };
  }
  if (typeof entry === "object" && entry.name) {
    const ref =
      entry.ref ??
      (entry.owner && entry.repo
        ? `${entry.owner}/${entry.repo}#${entry.name}`
        : entry.name);
    const normalized = normalizeDep(
      entry.version_range ? `${ref}@${entry.version_range}` : ref,
    );
    return normalized;
  }
  return null;
}

function normalizeAllowedTools(v) {
  if (v == null) return [];
  if (Array.isArray(v)) return v.map(String);
  if (typeof v === "string") {
    return v
      .split(",")
      .map((s) => s.trim())
      .filter(Boolean);
  }
  return [];
}

function normalizeMetadata(v) {
  if (v == null || typeof v !== "object") return {};
  const out = {};
  for (const [k, val] of Object.entries(v)) {
    out[String(k)] = val == null ? "" : String(val);
  }
  return out;
}

export function parseFrontmatter(text) {
  if (typeof text !== "string") {
    throw new MissingFrontmatterError("input is not a string");
  }
  // Allow optional BOM and leading whitespace before the opening delimiter.
  const stripped = text.replace(/^﻿/, "");
  const lines = stripped.split("\n");
  if (lines.length === 0 || lines[0].trim() !== "---") {
    throw new MissingFrontmatterError(
      "frontmatter must begin with '---' on the first line",
    );
  }
  let closeIdx = -1;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === "---") {
      closeIdx = i;
      break;
    }
  }
  if (closeIdx === -1) {
    throw new MissingFrontmatterError(
      "frontmatter has no closing '---' delimiter",
    );
  }
  const yamlBlock = lines.slice(1, closeIdx).join("\n");
  let data;
  try {
    data = parseYaml(yamlBlock) ?? {};
  } catch (e) {
    throw new MalformedFrontmatterError(
      `frontmatter YAML is malformed: ${e.message}`,
      e,
    );
  }
  if (typeof data !== "object" || Array.isArray(data)) {
    throw new MalformedFrontmatterError(
      "frontmatter YAML must be a mapping",
    );
  }
  const body = lines.slice(closeIdx + 1).join("\n");
  const source_offset =
    lines.slice(0, closeIdx + 1).join("\n").length + 1; // newline after closing ---
  const deps = Array.isArray(data.dependencies)
    ? data.dependencies.map(normalizeDep).filter(Boolean)
    : [];
  return {
    name: data.name != null ? String(data.name) : null,
    description: data.description != null ? String(data.description) : null,
    tier: data.tier != null ? String(data.tier) : null,
    license: data.license != null ? String(data.license) : null,
    compatibility:
      data.compatibility != null ? String(data.compatibility) : null,
    metadata: normalizeMetadata(data.metadata),
    allowed_tools: normalizeAllowedTools(data["allowed-tools"] ?? data.allowed_tools),
    dependencies: deps,
    body,
    source_offset,
    raw_frontmatter: data,
  };
}

export function toCanonicalJson(parsed) {
  return JSON.stringify(sortKeys(parsed));
}

function sortKeys(v) {
  if (Array.isArray(v)) return v.map(sortKeys);
  if (v && typeof v === "object") {
    const out = {};
    for (const k of Object.keys(v).sort()) out[k] = sortKeys(v[k]);
    return out;
  }
  return v;
}

File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/parser.py
from __future__ import annotations

import io
import re
from typing import Optional

from ruamel.yaml import YAML
from ruamel.yaml.error import YAMLError

from .types import DepRef, Parsed


class MissingFrontmatterError(Exception):
    pass


class MalformedFrontmatterError(Exception):
    pass


_DEP_RE = re.compile(
    r"^(?:(?P<owner>[A-Za-z0-9_.-]+)/(?P<repo>[A-Za-z0-9_.-]+)\#)?"
    r"(?P<name>[a-z0-9-]+)(?:@(?P<range>.+))?$"
)


def _normalize_dep(entry) -> Optional[DepRef]:
    if entry is None:
        return None
    if isinstance(entry, str):
        m = _DEP_RE.match(entry.strip())
        if not m:
            return DepRef(
                name=entry.strip(),
                version_range=None,
                origin="in_repo",
                owner=None,
                repo=None,
            )
        owner = m.group("owner")
        return DepRef(
            name=m.group("name"),
            version_range=m.group("range"),
            origin="cross_repo" if owner else "in_repo",
            owner=owner,
            repo=m.group("repo"),
        )
    if isinstance(entry, dict) and entry.get("name"):
        ref = entry.get("ref")
        if not ref:
            if entry.get("owner") and entry.get("repo"):
                ref = f"{entry['owner']}/{entry['repo']}#{entry['name']}"
            else:
                ref = entry["name"]
        if entry.get("version_range"):
            ref = f"{ref}@{entry['version_range']}"
        return _normalize_dep(ref)
    return None


def _normalize_allowed_tools(v) -> list:
    if v is None:
        return []
    if isinstance(v, list):
        return [str(x) for x in v]
    if isinstance(v, str):
        return [s.strip() for s in v.split(",") if s.strip()]
    return []


def _normalize_metadata(v) -> dict:
    if not isinstance(v, dict):
        return {}
    return {str(k): "" if val is None else str(val) for k, val in v.items()}


def parse_frontmatter(text: str) -> Parsed:
    if not isinstance(text, str):
        raise MissingFrontmatterError("input is not a string")
    stripped = text.lstrip("﻿")
    lines = stripped.split("\n")
    if not lines or lines[0].strip() != "---":
        raise MissingFrontmatterError(
            "frontmatter must begin with '---' on the first line"
        )
    close_idx = -1
    for i in range(1, len(lines)):
        if lines[i].strip() == "---":
            close_idx = i
            break
    if close_idx == -1:
        raise MissingFrontmatterError(
            "frontmatter has no closing '---' delimiter"
        )
    yaml_block = "\n".join(lines[1:close_idx])
    yaml_loader = YAML(typ="safe")
    try:
        data = yaml_loader.load(io.StringIO(yaml_block))
    except YAMLError as e:
        raise MalformedFrontmatterError(
            f"frontmatter YAML is malformed: {e}"
        ) from e
    if data is None:
        data = {}
    if not isinstance(data, dict):
        raise MalformedFrontmatterError("frontmatter YAML must be a mapping")
    body = "\n".join(lines[close_idx + 1 :])
    source_offset = len("\n".join(lines[: close_idx + 1])) + 1
    raw_deps = data.get("dependencies") or []
    deps: list = []
    if isinstance(raw_deps, list):
        for e in raw_deps:
            d = _normalize_dep(e)
            if d is not None:
                deps.append(d)
    return Parsed(
        name=str(data["name"]) if data.get("name") is not None else None,
        description=str(data["description"])
        if data.get("description") is not None
        else None,
        tier=str(data["tier"]) if data.get("tier") is not None else None,
        license=str(data["license"]) if data.get("license") is not None else None,
        compatibility=str(data["compatibility"])
        if data.get("compatibility") is not None
        else None,
        metadata=_normalize_metadata(data.get("metadata")),
        allowed_tools=_normalize_allowed_tools(
            data.get("allowed-tools") if "allowed-tools" in data else data.get("allowed_tools")
        ),
        dependencies=deps,
        body=body,
        source_offset=source_offset,
        raw_frontmatter=dict(data),
    )


def to_canonical_json(obj) -> str:
    import json

    def to_plain(v):
        if hasattr(v, "to_dict"):
            return to_plain(v.to_dict())
        if isinstance(v, list):
            return [to_plain(x) for x in v]
        if isinstance(v, dict):
            return {str(k): to_plain(val) for k, val in v.items()}
        return v

    return json.dumps(to_plain(obj), sort_keys=True, separators=(",", ":"))
